
#+TITLE: KmerDB (.kdb)  [TURN THAT TSHI UP]
#+SUBTITLE: A simple bgzf-based k-mer database format
#+AUTHOR: Matthew Ralston <mralston.development@gmail.com>

# .kdb files should be debrujin graph databases
# The final prototype would be .bgzf format from biopython
* 6/14/25 | [0.8.17] codons added. kmer, parse module fixed
Released on PyPI as 0.8.17
Fixed kmer.py module kmer_to_id and id_to_kmer functions, added is_aa amino acid conversions
parse.py module: correct min/max sequence/read lengths added to header. Was putting placeholder 50,0 min max sequence lengths before.
Adds codons.codon_frequency_table() which returns a codon list, codon counts, and freqencies
Adds codons.get_codons_in_order which is a list of 3-mer ids
Both functions make sure L%3 == 0

